models/helpers.ts

```diff
@@ -130,6 +133,142 @@ export function getLedgerLink(
   };
 }
 
+function getMakeCreditNoteAction(
+  fyo: Fyo,
+  schemaName: ModelNameEnum.SalesInvoice | ModelNameEnum.PurchaseInvoice
+): Action {
+  return {
+    label: fyo.t`Return / Credit Note`,
+    group: fyo.t`Create`,
+    condition: (doc: Doc) =>
+      schemaName === ModelNameEnum.SalesInvoice &&
+      doc.isSubmitted &&
+      !doc.isReturn &&
+      !doc.returnCompleted,
+    action: async (doc: Doc) => {
+      const salesInvoice = doc.getValidDict(true, true) as DocValueMap;
+      const rawCreditNote = (await createReturnDoc(
+        salesInvoice,
+        ModelNameEnum.SalesInvoice,
+        fyo
+      )) as DocValueMap;
+
+      const rawCreditNoteDoc = doc.fyo.doc.getNewDoc(
+        doc.schemaName,
+        rawCreditNote,
+        true
+      ) as Invoice;
+
+      const { openEdit } = await import('src/utils/ui');
+      await openEdit(rawCreditNoteDoc);
+    },
+  };
+}
+export async function updateReturnCompleteStatus(doc: Invoice) {
+  if (!doc.isReturn || !doc.returnAgainst || !doc.items) {
+    return;
+  }
+
+  for (const item of doc.items) {
```

Comment:
Very inefficient function just for the purpose of checking whether all the items have been returned.

The `group by` version of `getInvoiceReturnedItemQty` can be checked against the items of the quantity of items in the doc for this.

That's two db calls, as opposed to `items.length * 2`.